cirq-core/cirq/testing/routing_devices.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Provides test devices that can validate circuits during a routing procedure."""
+
+from typing import Optional, TYPE_CHECKING
+import networkx as nx
+
+from cirq import devices
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class RoutingTestingDevice(devices.Device):
+    """Testing device to be used only for testing qubit connectivity in routing procedures."""
+
+    def __init__(self, metadata: devices.DeviceMetadata) -> None:
+        self._metadata = metadata
+
+    @property
+    def metadata(self) -> Optional[devices.DeviceMetadata]:
+        return self._metadata
+
+    def validate_operation(self, operation: 'cirq.Operation') -> None:
+        for q in operation.qubits:
+            if q not in self._metadata.qubit_set:
+                raise ValueError(f'Qubit not on device: {q!r}.')
+
+        if (
+            len(operation.qubits) == 2
+            and operation.qubits[0] not in self._metadata.nx_graph[operation.qubits[1]]
+        ):
+            raise ValueError(f'Qubit pair is not valid on device: {operation.qubits!r}.')
+
+
+def construct_grid_device(d: int) -> RoutingTestingDevice:
```

Comment:
function name is inconsistent with return type's name

Reply:
At some point I want to implement a function `construct_ring_device()` and `construct_butterfly_device()` which will all be used for testing routing procedures. What do you recommend I do?

Reply:
We should consider changing the constructor of `RoutingTestingDevice` to expect a graph on integers and then automatically:

1. Relabel nodes using `cirq.NamedQubit(node) for node in input_graph`.
2. Construct a device metadata object using the device graph and qubits generated above. 

Then, these functions would just be `construct_grid_graph` / `construct_ring_graph` etc., which can be done using direct function calls to networkx.

 See https://networkx.org/documentation/stable/reference/generators.html for reference.

Reply:
> At some point I want to implement a function construct_ring_device() and construct_butterfly_device() which will all be used for testing routing procedures.

ok that makes sense

Reply:
re: @tanujkhattar 's suggestion. Yes, it's kindof shocking that we still don't have a `GraphDevice` that just takes a graph. This could be it! (might be worth moving out of the testing module in this case)

Using `NamedQubits` strikes me as slightly strange, but something is necessary to translate between the networkx interface (which allows anything to be nodes) to cirq's design (which only allows Qid's to be in Devices). Another option would be to try to translate to the appropriate type:

```python
if is_tuple_of_two_numbers(x):
  node = cirq.GridQubit(*x)
if is_one_integer(x):
  node = cirq.LineQubit(x)
else:
  node = cirq.NamedQubit(str(x))
```

might be a bit too "magic" for some folks' taste; but could be very helpful in practice

Reply:
I guess `DeviceMetadata` already just takes a graph, and this class just takes a Metadata  -- so it's pretty close. It still shocks me that we have to write this class and there isn't a device that just delegates to its metadata which is just initialized with a graph

Reply:
This sounds like a good idea to me.